### PR TITLE
Extend CosineDist support for number types other than Float64

### LINKED ISSUE #55: `CosineDist` fails with Ints
Just ran into this:

``` julia
julia> evaluate(CosineDist(),[1,0],[1,0])
ERROR: BoundsError
 in indexed_next at ./tuple.jl:35 [inlined]
 in eval_reduce(::Distances.CosineDist, ::Float64, ::Tuple{Int64,Int64,Int64}) at /Users/alex/.julia/v0.5/Distances/src/metrics.jl:140
 in macro expansion at /Users/alex/.julia/v0.5/Distances/src/metrics.jl:74 [inlined]
 in macro expansion at ./simdloop.jl:73 [inlined]
 in evaluate(::Distances.CosineDist, ::Array{Int64,1}, ::Array{Int64,1}) at /Users/alex/.julia/v0.5/Distances/src/metrics.jl:71
```

It works if the vectors contain floats:

``` julia
julia> evaluate(CosineDist(),[1,0.],[1,0.])
0.0
```

### PR DESCRIPTION
**Pull request by ntdef**
Fixes #55.

### CODE REVIEW

File: src/metrics.jl [hunk 1]
```diff
@@ -132,7 +132,7 @@ hamming(a::AbstractArray, b::AbstractArray) = evaluate(Hamming(), a, b)
 hamming{T <: Number}(a::T, b::T) = evaluate(Hamming(), a, b)
 
 # Cosine dist
-function eval_start{T<:AbstractFloat}(::CosineDist, a::AbstractArray{T}, b::AbstractArray{T})
+function eval_start{T<:Number}(::CosineDist, a::AbstractArray{T}, b::AbstractArray{T})
```
Comment (ararslan) [inline src/metrics.jl]: Should probably be `Real` instead of `Number`. I'm not sure that the cosine distance is actually meaningful for complex numbers.

**Comment by ararslan** [comment]
It would also be great if you could add a test or two to verify that it works with integers.

**Comment by ntdef** [comment]
I changed the parent type and added unit tests. Tests passed locally.

**Comment by ararslan** [comment]
Thanks!

**Pull request merged**